boot/assets.go

```diff
@@ -289,14 +365,161 @@ type TrustedAssetsUpdateObserver struct{}
 //
 // Implements gadget.ContentUpdateObserver.
 func (o *TrustedAssetsUpdateObserver) Observe(op gadget.ContentOperation, affectedStruct *gadget.LaidOutStructure, root, realSource, relativeTarget string) (bool, error) {
+	var whichBootloader bootloader.Bootloader
+	var whichAssets []string
+	var err error
+	var isRecovery bool
+
+	switch affectedStruct.Role {
+	case gadget.SystemBoot:
+		if o.bootBootloader == nil {
+			o.bootBootloader, o.bootTrustedAssets, err = findTrustedAssetsBootloader(root, &bootloader.Options{
+				NoSlashBoot: true,
+			})
+			if err != nil {
+				return false, err
+			}
+		}
+		whichBootloader = o.bootBootloader
+		whichAssets = o.bootTrustedAssets
+	case gadget.SystemSeed:
+		if o.seedBootloader == nil {
+			o.seedBootloader, o.seedTrustedAssets, err = findTrustedAssetsBootloader(root, &bootloader.Options{
+				NoSlashBoot: true,
+				Recovery:    true,
+			})
+			if err != nil {
+				return false, err
+			}
+		}
+		whichBootloader = o.seedBootloader
+		whichAssets = o.seedTrustedAssets
+		isRecovery = true
+	default:
+		// only system-seed and system-boot are of interest
+		return true, nil
+	}
+	if len(whichAssets) == 0 || !strutil.ListContains(whichAssets, relativeTarget) {
+		// not one of the trusted assets
+		return true, nil
+	}
+	if o.modeenv == nil {
+		// we've hit a trusted asset, so a modeenv is needed now too
+		o.modeenv, err = ReadModeenv("")
+		if err != nil {
+			return false, fmt.Errorf("cannot load modeenv: %v", err)
+		}
+	}
+	switch op {
+	case gadget.ContentUpdate:
+		return o.observeUpdate(whichBootloader, isRecovery, root, realSource, relativeTarget)
+	case gadget.ContentRollback:
+		return o.observeRollback(whichBootloader, isRecovery, root, realSource, relativeTarget)
+	default:
+		// we only care about update and rollback actions
+		return false, nil
+	}
+}
+
+func (o *TrustedAssetsUpdateObserver) observeUpdate(bl bootloader.Bootloader, recovery bool, root, realSource, relativeTarget string) (bool, error) {
+	modeenvBefore, err := o.modeenv.Copy()
+	if err != nil {
+		return false, fmt.Errorf("cannot copy modeenv: %v", err)
+	}
+
+	ta, err := o.cache.Add(realSource, bl.Name(), filepath.Base(relativeTarget))
+	if err != nil {
+		return false, err
+	}
+
+	trustedAssets := &o.modeenv.CurrentTrustedBootAssets
+	if recovery {
+		trustedAssets = &o.modeenv.CurrentTrustedRecoveryBootAssets
+	}
+	if !isAlreadyTrackedInBootMap(*trustedAssets, ta) {
+		if *trustedAssets == nil {
+			*trustedAssets = bootAssetsMap{}
+		}
+		(*trustedAssets)[ta.name] = append((*trustedAssets)[ta.name], ta.hash)
```

Comment:
`isAlreadyTrackedInBootMap` checks whether `(asset name, hash)` is already tracked. If last time we reached a point where modeenv was written to, an entry for the same file will not be added again. If modedenv wasn't yet written, but the file is in cachec, the check will ensure that it gets added to modeenv too.

Reply:
Suppose a scenario where the modeenv was written, but some time passes until the next attempt and a new version of the asset is now available with a different hash. Is the previous operation resumed before updating again to the current version?

Reply:
There are two things to keep in mind here:

* there can be only one gadget update change active at any point (we'll need to be careful once kernels can also trigger gadget updates though, cc @mvo5 ) , a new one cannot start if the old one didn't finish or abort.  We need to be careful though, what happens on error cases, will we leave dangling hashes in modeenv ?

* on reboot there will be a MarkBootSuccessful that needs to deal again with the modeenv and reseal, like for kernels it will need to make sure that there is one hash (the effectively current one) left in the modeenv for each asset



*

Reply:
Dangling hashes seem bad as they can affect the transition reseal during the update of a different package. Ideally we should have it always consistent, if an error happens the hashes should revert to the previous state before any other operation takes place.

Reply:
I think we can add a simple sanity check to verify whether the asset is already on disk, and if so there's only one hash listed so far. If the file does not exist on disk yet, the asset should not yet be present in the map (or at least no hashes should be listed there).

Reply:
to be clear we don't want more errors,  we want not be sure we don't leave hashes in the map without a chance of them becoming the real assets by writing+reboot. I suppose Cancel plays a role here, but is that enough?